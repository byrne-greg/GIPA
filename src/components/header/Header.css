.RobotImage {
  animation-name: RobotImage_horizontal_flip;
  animation-duration: 2s;
  animation-iteration-count: infinite;
  animation-direction: alternate;
  
  height: calc(45px + 2vw);
  margin: 0 20px;
}

@keyframes RobotImage_horizontal_flip {
  from { -webkit-transform: scaleX(0); transform: scaleX(0); }
  to { -webkit-transform: scaleX(1); transform: scaleX(1); }
}

.Header {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;

  padding: 15px;
  margin: 0 0 10px 0;
  border-bottom: 1px solid black;

  color: black;
  background-color: #e5e5e5;
}

.HeaderTitleContainer {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
}

.HeaderTitleText {
  display: flex;
}

.HeaderTitleChar {
  margin: 0 3px;
}

.HeaderSubTitle {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
}

.HeaderSubTitle_title {
  margin: 0 5px 5px 5px;
}

.HeaderSubTitle_iconLinks {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
}

.HeaderSubTitle_iconLinks a {
  margin: 10px;
}

.fab {
  padding: 5px;
  font-size: 20px;
  width: 30px;
  text-align: center;
  text-decoration: none;
  margin: 5px 2px;
  border-radius: 50%;
}

.fab:hover {
    opacity: 0.7;
}

.fa-facebook-f {
  background: #3B5998;
  color: white;
}

.fa-linkedin-in {
  background: #007bb5;
  color: white;
}

.fa-github {
  background: black;
  color: white;
}





