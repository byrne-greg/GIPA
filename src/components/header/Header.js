import "./Header.css";
import React from "react";
import PropTypes from "prop-types";
import shortid from "shortid";

export default class Header extends React.Component {
  // default state
  state = {
    title: "GIPA",
    subTitle: "By Greg Byrne"
  };

  render() {
    return (
        <header className="Header">
          <HeaderTitleContainer title={this.state.title}/>
          <HeaderSubTitle subTitle={this.state.subTitle}/>
        </header>
    );
  }
}

export const HeaderTitleContainer = (props) => {
  return(
    <div className="HeaderTitleContainer">
      <RobotImage/>
      <HeaderTitleText titlechars={props.title.split("")}/>
      <RobotImage/>
    </div>
  );
}
HeaderTitleContainer.propTypes = {
  title: PropTypes.string.isRequired
};
HeaderTitleContainer.defaultProps = {
  title: "GIPA"
};

export const RobotImage = () => {
  return(
    <img src="https://upload.wikimedia.org/wikipedia/commons/thumb/7/7f/Robot.svg/1024px-Robot.svg.png" className="RobotImage" alt="Robot. Icon made by Freepik from www.flaticon.com " />
  );
}

export const HeaderTitleText = (props) => {
  return(
    <h1 className="HeaderTitleText">
      {props.titlechars.map(char => <HeaderTitleChar key={shortid.generate()} char={char}/>)}
    </h1>
  );
}
HeaderTitleText.propTypes = {
  titlechars: PropTypes.arrayOf(PropTypes.string.isRequired)
};
HeaderTitleText.defaultProps = {
  titlechars: ["G","I","P","A"]
};

export const HeaderTitleChar = (props) => {
  return(
    <div className="HeaderTitleChar">{props.char}</div>
  );
}
HeaderTitleChar.propTypes = {
  char: PropTypes.string
};

export const HeaderSubTitle = (props) => {
  return(
    <div className="HeaderSubTitle">
      <h3 className="HeaderSubTitle_title">{props.subTitle}</h3>
      <div className="HeaderSubTitle_iconLinks">
        <IconLink href="https://github.com/byrne-greg" iconClassName="fab fa-github" screenReaderText="Link: Greg Byrne Github Profile"/>
        <IconLink href="https://www.linkedin.com/in/byrne-greg/" iconClassName="fab fa-linkedin-in" screenReaderText="Link: Greg Byrne LinkedIn Profile"/>
        <IconLink href="https://www.facebook.com/greg.byrne.3551" iconClassName="fab fa-facebook-f" screenReaderText="Link: Greg Byrne Facebook Profile"/>
      </div>
    </div>
  );
}
HeaderSubTitle.propTypes = {
  subTitle: PropTypes.string
};
HeaderSubTitle.defaultProps = {
  subTitle: "By Greg Byrne"
};

export const IconLink = (props) => {
  return(
    <a href={props.href} target="_blank" rel="noopener noreferrer" className={props.iconClassName}><span className="hidden">{props.screenReaderText}</span></a>
  );
}
IconLink.propTypes = {
  href: PropTypes.string.isRequired,
  iconClassName: PropTypes.string.isRequired,
  screenReaderText: PropTypes.string.isRequired
};
IconLink.defaultProps = {
  href: "#",
  iconClassName: "far fa-frown",
  screenReaderText: "Default Icon Link"
};


