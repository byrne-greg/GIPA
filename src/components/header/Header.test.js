import React from "react";
import Header, {HeaderTitleContainer, RobotImage, HeaderTitleText, HeaderSubTitle, HeaderTitleChar, IconLink } from "./Header";
import renderer from "react-test-renderer";
import {shallow, mount} from "enzyme";

describe("Header component", () => {
 
	it("shallow renders without crashing", () => {
		shallow(<Header/>);
	});
  
	it("shallow renders with the correct number/type of children components", () => {
		var wrapper = shallow(<Header/>);
		expect(wrapper.find(HeaderTitleContainer).length).toBe(1);
		expect(wrapper.find(HeaderSubTitle).length).toBe(1);
	});

	it("full render matches expected snapshot", () => {
		const header = renderer.create(<Header/>).toJSON();
		expect(header).toMatchSnapshot();
	});
});

describe("HeaderTitleContainer functional component", () => {

	it("shallow renders without crashing", () => {
		shallow(<HeaderTitleContainer/>);
    
	});

	it("shallow renders with the correct number/type of children components", () => {
		var wrapper = shallow(<HeaderTitleContainer/>);
		expect(wrapper.find(RobotImage).length).toBe(2);
		expect(wrapper.find(HeaderTitleText).length).toBe(1);
	});

	it("shallow renders and passes the correct prop format to HeaderTitleText", () => {
		var wrapper = shallow(<HeaderTitleContainer title="test"/>);
		expect(wrapper.find(HeaderTitleText).prop("titlechars")).toEqual(["t","e","s","t"]);
	});

	it("shallow renders and passes the correct default prop format to HeaderTitleText", () => {
		var wrapper = shallow(<HeaderTitleContainer/>);
		expect(wrapper.find(HeaderTitleText).prop("titlechars")).toEqual(["G","I","P","A"]);
	});

});

describe("RobotImage functional component", () => {

	it("shallow renders without crashing", () => {
		shallow(<RobotImage/>);
	});

});

describe("HeaderTitleText functional component", () => {

	it("shallow renders without crashing", () => {
		shallow(<HeaderTitleText/>);
	});

	it("renders with correct amount of children per prop", () => {
		var wrapper = mount(<HeaderTitleText titlechars={["1","2"]}/>);
		expect(wrapper.find(HeaderTitleChar).length).toBe(2);
	});

	it("renders with correct amount of children per default prop", () => {
		var wrapper = mount(<HeaderTitleText />);
		expect(wrapper.find(HeaderTitleChar).length).toBe(4);
	});

});

describe("HeaderTitleChar functional component", () => {

	it("shallow renders without crashing without prop", () => {
		shallow(<HeaderTitleChar />);
	});

	it("shallow renders without crashing with prop", () => {
		shallow(<HeaderTitleChar char="c" />);
	});

});

describe("HeaderSubTitle functional component", () => {

	it("shallow renders without crashing", () => {
		shallow(<HeaderSubTitle/>);
	});

	it("shallow renders with default subtitle prop", () => {
		var wrapper = shallow(<HeaderSubTitle/>);
		expect(wrapper.find(".HeaderSubTitle_title").text()).toEqual("By Greg Byrne");
	});

	it("shallow renders with subtitle prop", () => {
		var wrapper = shallow(<HeaderSubTitle subTitle="test"/>);
		expect(wrapper.find(".HeaderSubTitle_title").text()).toEqual("test");
	});

	it("shallow renders with correct number of IconLink components", () => {
		var wrapper = shallow(<HeaderSubTitle/>);
		expect(wrapper.find(IconLink).length).toBe(3);
	});

});

describe("IconLink functional component", () => {

	it("shallow renders without crashing", () => {
		shallow(<IconLink/>);
	});

	it("shallow renders with correct props", () => {
		var wrapper = shallow(<IconLink href="test href" screenReaderText="test screenReaderText" iconClassName="test iconClassName"/>);
		expect(wrapper.find("a").prop("href")).toEqual("test href");
		expect(wrapper.find(".hidden").text()).toEqual("test screenReaderText");
		expect(wrapper.find("a").prop("className")).toEqual("test iconClassName");
	});

});
