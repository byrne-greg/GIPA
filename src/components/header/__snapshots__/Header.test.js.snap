// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component full render matches expected snapshot 1`] = `
<header
  className="Header"
>
  <div
    className="HeaderTitleContainer"
  >
    <img
      alt="Robot. Icon made by Freepik from www.flaticon.com "
      className="RobotImage"
      src="https://upload.wikimedia.org/wikipedia/commons/thumb/7/7f/Robot.svg/1024px-Robot.svg.png"
    />
    <h1
      className="HeaderTitleText"
    >
      <div
        className="HeaderTitleChar"
      >
        G
      </div>
      <div
        className="HeaderTitleChar"
      >
        I
      </div>
      <div
        className="HeaderTitleChar"
      >
        P
      </div>
      <div
        className="HeaderTitleChar"
      >
        A
      </div>
    </h1>
    <img
      alt="Robot. Icon made by Freepik from www.flaticon.com "
      className="RobotImage"
      src="https://upload.wikimedia.org/wikipedia/commons/thumb/7/7f/Robot.svg/1024px-Robot.svg.png"
    />
  </div>
  <div
    className="HeaderSubTitle"
  >
    <h3
      className="HeaderSubTitle_title"
    >
      By Greg Byrne
    </h3>
    <div
      className="HeaderSubTitle_iconLinks"
    >
      <a
        className="fab fa-github"
        href="https://github.com/byrne-greg"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span
          className="hidden"
        >
          Link: Greg Byrne Github Profile
        </span>
      </a>
      <a
        className="fab fa-linkedin-in"
        href="https://www.linkedin.com/in/byrne-greg/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span
          className="hidden"
        >
          Link: Greg Byrne LinkedIn Profile
        </span>
      </a>
      <a
        className="fab fa-facebook-f"
        href="https://www.facebook.com/greg.byrne.3551"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span
          className="hidden"
        >
          Link: Greg Byrne Facebook Profile
        </span>
      </a>
    </div>
  </div>
</header>
`;
